mobile/test/java/org/chromium/net/CronetHttp3Test.java

```diff
@@ -83,57 +90,212 @@ public void log(int logLevel, String message) {
+
+    // Make sure the request succeeded. It should go out over HTTP/2 as it's the first
+    // request and HTTP/3 support is not established.
+    assertEquals(200, callback.mResponseInfo.getHttpStatusCode());
+    assertEquals("h2", callback.mResponseInfo.getNegotiatedProtocol());
   }
 
   @Test
   @SmallTest
   @Feature({"Cronet"})
-  public void testInitEngineAndStartRequest() throws Exception {
+  public void basicHttp3Get() throws Exception {
     // Ideally we could override this from the command line but that's TBD.
     setUp(printEnvoyLogs);
 
-    // Set up the Envoy engine.
-    NativeCronvoyEngineBuilderImpl nativeCronetEngineBuilder =
-        new NativeCronvoyEngineBuilderImpl(ApplicationProvider.getApplicationContext());
-    if (printEnvoyLogs) {
-      nativeCronetEngineBuilder.setLogger(logger);
-      nativeCronetEngineBuilder.setLogLevel(EnvoyEngine.LogLevel.TRACE);
-    }
-    // Make sure the handshake will work despite lack of real certs.
-    nativeCronetEngineBuilder.setMockCertVerifierForTesting();
-    cronvoyEngine = new CronvoyUrlRequestContext(nativeCronetEngineBuilder);
+    // Do the initial HTTP/2 request to get the alt-svc response.
+    doInitialHttp2Request();
 
-    // Do a request to https://127.0.0.1:test_server_port/
-    TestUrlRequestCallback callback1 = new TestUrlRequestCallback();
-    String newUrl = "https://" + http2TestServer.getAddress() + "/";
+    // Set up a second request, which will hopefully go out over HTTP/3 due to alt-svc
+    // advertisement.
+    TestUrlRequestCallback callback = doBasicGetRequest();
+
+    // Verify the second request used HTTP/3
+    assertEquals(200, callback.mResponseInfo.getHttpStatusCode());
+    assertEquals("h3", callback.mResponseInfo.getNegotiatedProtocol());
+  }
+
+  @Test
+  @SmallTest
+  @Feature({"Cronet"})
+  public void failToHttp2() throws Exception {
+    // Ideally we could override this from the command line but that's TBD.
+    setUp(printEnvoyLogs);
+
+    // Do the initial HTTP/2 request to get the alt-svc response.
+    doInitialHttp2Request();
+
+    // Set up a second request, which will hopefully go out over HTTP/3 due to alt-svc
+    // advertisement.
+    TestUrlRequestCallback get_callback = doBasicGetRequest();
+
+    // Verify the second request used HTTP/3
+    assertEquals(200, get_callback.mResponseInfo.getHttpStatusCode());
+    assertEquals("h3", get_callback.mResponseInfo.getNegotiatedProtocol());
+
+    // Now stop the HTTP/3 server.
+    http3TestServer.shutdown();
+    http3TestServer = null;
+
+    // The next request will fail on HTTP2 but should succeed on HTTP/2 despite having a body.
+    TestUrlRequestCallback post_callback = doBasicPostRequest();
+    assertEquals(200, post_callback.mResponseInfo.getHttpStatusCode());
+    assertEquals("h2", post_callback.mResponseInfo.getNegotiatedProtocol());
+  }
+
+  @Test
+  @SmallTest
+  @Feature({"Cronet"})
+  public void testNoRetryPostPostHandshake() throws Exception {
```

Comment:
nits: s/PostHandshake/AfterHandshakeCompletes. POST request v.s. post handshake is confusing.

Reply:
hah, fair point